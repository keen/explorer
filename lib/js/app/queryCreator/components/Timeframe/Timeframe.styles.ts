import styled from 'styled-components';
import { colors } from '@keen.io/colors';

export const Container = styled.div`
  position: relative;
`;

export const SettingsContainer = styled.div`
  padding: 15px 10px;
  border-top: solid 1px ${colors.white[300]};
  border-bottom: solid 1px ${colors.white[300]};
`;
