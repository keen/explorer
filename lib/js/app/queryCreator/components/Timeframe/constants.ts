import text from './text.json';

export const RELATIVE_TAB = 'relative';
export const ABSOLUTE_TAB = 'absolute';

export const TABS_SETTINGS = [
  {
    label: text.relative,
    id: RELATIVE_TAB,
  },
  {
    label: text.absolute,
    id: ABSOLUTE_TAB,
  },
];
