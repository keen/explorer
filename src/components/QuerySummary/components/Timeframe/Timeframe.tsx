import React, { FC } from 'react';

import { Timeframe, Timezones } from '@keen.io/query-creator';
import AbsoluteTimeframe from '../AbsoluteTimeframe';

type Props = {
  timeframe: Timeframe;
  timezone?: Timezones | number;
};

const Timeframe: FC<Props> = ({ timeframe, timezone }) =>
  typeof timeframe !== 'string' ? (
    <AbsoluteTimeframe timeframe={timeframe} timezone={timezone} />
  ) : (
    <span>{timeframe}</span>
  );

export default Timeframe;
