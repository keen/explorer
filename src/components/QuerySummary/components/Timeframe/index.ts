import Timeframe from './Timeframe';

export default Timeframe;
