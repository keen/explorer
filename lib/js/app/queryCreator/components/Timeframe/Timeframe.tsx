import React, { FC, useState, useRef, useEffect } from 'react';

import { Container, SettingsContainer } from './Timeframe.styles';
import text from './text.json';

import Title from '../Title';
import Tabs from '../Tabs';
import Dropdown from '../Dropdown';
import DropableContainer, { Variant } from '../DropableContainer';
import AbsoluteTime, { TIME_PICKER_CLASS } from '../AbsoluteTime';
import AbsoluteTimeLabel from '../AbsoluteTimeLabel';
import RelativeTime from '../RelativeTime';
import RelativeTimeLabel from '../RelativeTimeLabel';
import Timezone, { getTimezoneValue } from '../Timezone';

import { getDefaultAbsoluteTime } from './utils/getDefaultAbsoluteTime';
import { convertRelativeTime } from './utils/convertRelativeTime';

import { ABSOLUTE_TAB, RELATIVE_TAB, TABS_SETTINGS } from './constants';
import { DEFAULT_TIMEFRAME } from '../../modules/query';

import { Timeframe as TimeframeType, Timezones } from '../../types';

type Props = {
  /** Unique identifer */
  id: string;
  /** Timeframe change event handler */
  onTimeframeChange: (timeframe: TimeframeType) => void;
  /** Timezone change event handler */
  onTimezoneChange: (timezone: Timezones) => void;
  /** Timezone value */
  timezone: number | Timezones;
  /** Current timeframe value */
  value: TimeframeType;
  /** Reset field event handler */
  onReset?: () => void;
  /** Container variant */
  variant?: Variant;
};

const Timeframe: FC<Props> = ({
  id,
  onTimeframeChange,
  onTimezoneChange,
  onReset,
  timezone,
  value,
  variant = 'primary',
}) => {
  const [isOpen, setOpen] = useState(false);
  const containerRef = useRef(null);

  useEffect(() => {
    return () => {
      if (onReset) onReset();
    };
  }, []);

  const timezoneValue = getTimezoneValue(timezone);

  return (
    <Container ref={containerRef}>
      <Title onClick={() => !isOpen && setOpen(true)}>{text.label}</Title>
      <DropableContainer
        variant={variant}
        onClick={() => !isOpen && setOpen(true)}
        isActive={isOpen}
        value={value}
        onDefocus={(event: any) => {
          if (
            !event.path?.includes(containerRef.current) &&
            !event.path?.includes(
              document.querySelector(`.${TIME_PICKER_CLASS}`)
            )
          ) {
            setOpen(false);
          }
        }}
      >
        {typeof value === 'string' ? (
          <RelativeTimeLabel {...convertRelativeTime(value)} />
        ) : (
          <AbsoluteTimeLabel
            start={value.start}
            end={value.end}
            timezone={timezoneValue}
          />
        )}
      </DropableContainer>
      <Dropdown isOpen={isOpen}>
        <Tabs
          activeTab={typeof value === 'string' ? RELATIVE_TAB : ABSOLUTE_TAB}
          onClick={(tabId) => {
            if (tabId === RELATIVE_TAB) {
              onTimeframeChange(DEFAULT_TIMEFRAME);
            } else {
              onTimeframeChange(getDefaultAbsoluteTime(timezoneValue));
            }
          }}
          tabs={TABS_SETTINGS}
        />
        <SettingsContainer>
          {typeof value === 'string' ? (
            <RelativeTime
              onChange={onTimeframeChange}
              {...convertRelativeTime(value)}
            />
          ) : (
            <AbsoluteTime
              id={id}
              {...value}
              timezone={timezoneValue}
              onChange={onTimeframeChange}
            />
          )}
        </SettingsContainer>
        <Timezone
          timezone={timezoneValue}
          onChange={(timezone) => onTimezoneChange(timezone)}
        />
      </Dropdown>
    </Container>
  );
};

export default Timeframe;
