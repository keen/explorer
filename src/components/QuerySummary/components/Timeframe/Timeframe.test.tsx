import React from 'react';
import 'jest-styled-components';
import { render as rtlRender } from '@testing-library/react';

import Timeframe from './Timeframe';

const render = (overProps: any = {}) => {
  const props = {
    timeframe: 'this_14_days',
    timezone: 'UTC',
    ...overProps,
  };

  const wrapper = rtlRender(<Timeframe {...props} />);

  return {
    wrapper,
    props,
  };
};

test('should render relative timeframe', () => {
  const {
    wrapper: { getByText },
    props,
  } = render();

  const timeframe = getByText(props.timeframe);
  expect(timeframe).toBeInTheDocument();
});

test('should render absolute timeframe', () => {
  const {
    wrapper: { container },
  } = render({
    timeframe: { start: '2020-09-29T00:00:00Z', end: '2020-09-30T00:00:00Z' },
  });

  expect(container).toMatchSnapshot();
});
