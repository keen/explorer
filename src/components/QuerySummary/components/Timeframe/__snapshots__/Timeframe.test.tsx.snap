// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render absolute timeframe 1`] = `
.c0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c1 {
  margin: 0 5px;
  color: rgba(39,86,109,0.6);
}

<div>
  <div
    class="c0"
  >
    <span>
      2020-09-29 00:00
    </span>
    <span
      class="c1"
    >
      absolute_timeframe.separator
    </span>
    <span>
      2020-09-30 00:00
    </span>
  </div>
</div>
`;
